File: bb-plugin-token-usage-header/tools/threads_timeline.py
#!/usr/bin/env python3
"""Thread feed: the last N Claude Code sessions, laid out into time bins.

Reuses tools/tokens.py (walk(), Bucket, _session_files, _file_mtime,
JsonAwareParser) for the actual token counting — this module only handles
selecting the "last N sessions" and laying their usage records out into
fixed-size bins, and within a bin, by agent (main agent gets key "main",
subagents get their agentId, same as tokens.py --by agent).
"""
import json, os, sys
from collections import defaultdict
from datetime import datetime, timezone

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import tokens  # noqa: E402
import git_events  # noqa: E402

# Version of this script's --json report format. Separate from
# tokens.SCHEMA_VERSION — different contracts, different consumers
# (src/core/threads-timeline.ts checks this exact version). Bump on any
# breaking format change.
# 1 -> 2: added the top-level agentLabels field (human-readable agent names
# keyed by their key from bins) — the report shape expanded.
# 2 -> 3: a thread gained totalCost (usage cost in USD, priced per
# tokens.py) and workflowCount (number of distinct workflow runs in the
# session).
# 3 -> 4: a bin's workflow segment (agents[].key == "workflow:<run>", only
# under --group-workflows) gained members — a sorted list of the real
# agentIds merged into this segment; a regular (non-workflow) agent has no
# such field.
#
# 4 -> 5: a thread gained cwd/gitBranch (its working directory and git
# branch, for the chart's commit-marker enrichment on the TS side — see
# git_events.scan_session) and events (pr/push facts mined from its own
# transcript by the same function; commit events are appended later, by
# src/service/threads-timeline-service.ts, not by this script).
SCHEMA_VERSION = 5

# Truncation threshold for meta.description in agentLabels — the legend and
# labels in the UI aren't elastic, a long task description would break the
# chip/tooltip layout.
LABEL_DESCRIPTION_LIMIT = 60


def _session_main_files(root, project=None, session=None):
    """Main session .jsonl files (no subagents) across project directories.

    Returns a list of (sessionId, project, path, mtime) — one record per
    session, so "the last N by activity" can be selected BEFORE reading
    file contents. mtime is an approximation of "the transcript's last
    activity" (the same assumption as in tokens.py::_filter_by_since).
    """
    if not os.path.isdir(root):
        return []
    try:
        entries = os.listdir(root)
    except OSError:
        return []
    project_dirs = [os.path.join(root, d) for d in entries if os.path.isdir(os.path.join(root, d))]
    if project:
        project_dirs = [d for d in project_dirs if project in os.path.basename(d)]

    out = []
    for pd in project_dirs:
        try:
            names = os.listdir(pd)
        except OSError:
            continue
        for name in names:
            if not name.endswith(".jsonl"):
                continue
            path = os.path.join(pd, name)
            if not os.path.isfile(path):
                continue
            sid = name[: -len(".jsonl")]
            if session is not None and sid != session:
                continue
            mtime = tokens._file_mtime(path)
            out.append((sid, os.path.basename(pd), path, mtime.timestamp() if mtime else 0))
    return out


def _bin_start_epoch(epoch, unit):
    return (int(epoch) // unit) * unit


def _epoch(ts):
    """A record's timestamp (see tokens.walk) -> unix seconds, or None.

    Lenient parsing (as in walk() itself): a malformed/missing timestamp is
    not a user error, the record simply doesn't land in any bin.
    """
    dt = tokens._parse_time_bound(ts)
    return dt.timestamp() if dt else None


def _iso_from_epoch(epoch):
    """unix seconds -> ISO 8601 UTC with milliseconds. Same format as tokens.iso()."""
    dt = datetime.fromtimestamp(epoch, tz=timezone.utc)
    return dt.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"


def _truncate(s, limit=LABEL_DESCRIPTION_LIMIT):
    s = s.strip()
    if len(s) <= limit:
        return s
    return s[:limit].rstrip() + "…"


def _agent_label(agent_key, meta):
    """Human-readable agent label for agentLabels.

    "main" -> the fixed label "Main agent", not tied to meta (the main
    agent's meta is always None, see tokens.walk). For a subagent —
    meta.description (truncated), else meta.agentType, else the key itself
    as a fallback — the same priority ladder as agent_timeline.py::_header.
    """
    if agent_key == "main":
        return "Main agent"
    if meta:
        desc = meta.get("description")
        if desc:
            return _truncate(desc)
        atype = meta.get("agentType")
        if atype:
            return atype
    return agent_key


def _workflow_name(root, project, session, run_id):
    """Human-readable workflow name for its run_id.

    The run's script lives in `<project>/<session>/workflows/scripts/` and
    is named `<name>-<run_id>.js` (see the actual transcript layout) — the
    name is obtained by stripping the `-<run_id>.js` suffix. If the script
    is missing/the directory is unavailable — the run_id itself as a
    fallback (better to show the id than nothing).
    """
    scripts_dir = os.path.join(root, project, session, "workflows", "scripts")
    suffix = "-" + run_id + ".js"
    try:
        for name in os.listdir(scripts_dir):
            if name.endswith(suffix):
                return name[: -len(suffix)]
    except OSError:
        pass
    return run_id


def _bin_key(rec, group_workflows):
    """Bin segment key for a record.

    By default — the subagent's agentId ("main" for the main agent). Under
    group_workflows, all agents of one workflow run merge into a single
    `workflow:<run>` segment — so on the session page, a group of agents
    raised by one Workflow renders as a single segment (see agentLabels:
    there the key carries the workflow's name).
    """
    wf = rec["workflowRunId"]
    if group_workflows and wf:
        return "workflow:" + wf
    return rec["agentId"] or "main"


def build_timeline(root, limit=20, unit=300, project=None, session=None, group_workflows=False):
    """Builds the feed of the last `limit` sessions, laid out into `unit`-second bins.

    Doesn't redefine the usage-counting logic — every usage record from
    tokens.walk() is counted exactly once (dedup is already done in
    walk()), so the sum of total over a session's bins equals its full
    usage.

    Returns {"threads": [...], "agentLabels": {...}} — agentLabels is
    top-level (not per-thread): the same agentId across different threads
    (e.g. a repeated call to the same workflow agent) gets a single label.
    """
    sessions = _session_main_files(root, project=project, session=session)
    # the last N sessions by the transcript's last activity, freshest first
    sessions.sort(key=lambda row: row[3], reverse=True)
    sessions = sessions[: max(limit, 0)]

    project_dirs = {}
    # session id -> project dir basename (under `root`). Needed for
    # _workflow_name: rec["project"] from walk() is a relpath from the
    # global tokens.ROOT, not from the passed-in root, and it's wrong when
    # root != ROOT (tests).
    sid_to_proj = {}
    for _sid, proj, _path, _mtime in sessions:
        project_dirs.setdefault(proj, os.path.join(root, proj))
        sid_to_proj[_sid] = proj

    files = []
    # Kept per-session (not just flattened into `files`) so git_events.scan_session
    # can be run once per session below — it needs exactly one session's own
    # files, not the whole slice's.
    session_files = {}
    for sid, proj, _path, _mtime in sessions:
        sfiles = tokens._session_files(project_dirs[proj], sid)
        session_files[sid] = sfiles
        files.extend(sfiles)

    records_by_session = defaultdict(list)
    # agent_key -> label, populated from the meta of the FIRST record seen for
    # that key (across all sessions in this slice) — meta is the same for
    # every record of a given agent-<id>.jsonl file (one meta.json per agent
    # invocation), so which record "wins" doesn't change the label.
    agent_labels = {}
    for rec in tokens.walk(files):
        records_by_session[rec["session"]].append(rec)
        key = _bin_key(rec, group_workflows)
        if key not in agent_labels:
            if group_workflows and rec["workflowRunId"]:
                proj = sid_to_proj.get(rec["session"], rec["project"])
                agent_labels[key] = _workflow_name(root, proj, rec["session"], rec["workflowRunId"])
            else:
                agent_labels[key] = _agent_label(key, rec["meta"])

    threads = []
    for sid, proj, _path, _mtime in sessions:
        recs = records_by_session.get(sid, [])

        # bin_epoch -> agent_key -> Bucket. Bucket is reused from tokens.py,
        # so the total formula (inp+cacheWrite+cacheRead+out) isn't
        # duplicated and doesn't drift from the one tokens.py already applies.
        bin_buckets = defaultdict(lambda: defaultdict(tokens.Bucket))
        # bin_epoch -> workflow agent_key -> a set of the real agentIds
        # merged into this segment (see _bin_key: under group_workflows all
        # agents of one run collapse into "workflow:<run>," and the real
        # membership would otherwise be lost entirely — the frontend
        # couldn't tell whether a selected agent participated in this
        # segment, for highlighting/dimming on the session chart).
        # Populated only for workflow keys — for a regular agent, key is
        # already its real id, nothing to duplicate.
        bin_members = defaultdict(lambda: defaultdict(set))
        epochs = []
        for rec in recs:
            epoch = _epoch(rec["ts"])
            if epoch is None:
                continue
            epochs.append(epoch)
            bin_epoch = _bin_start_epoch(epoch, unit)
            agent_key = _bin_key(rec, group_workflows)
            bin_buckets[bin_epoch][agent_key].add(rec["usage"], rec["model"], rec["ts"])
            if agent_key.startswith("workflow:"):
                bin_members[bin_epoch][agent_key].add(rec["agentId"] or "main")

        if not epochs:
            # A session without a single valid usage record carries no data
            # for the feed (nothing to compute start/end/duration from) —
            # it's skipped, rather than emitted with fake zeros.
            continue

        start_epoch = min(epochs)
        end_epoch = max(epochs)

        # Continuous feed: one bin per EVERY time unit of the [start, end]
        # interval, including empty ones (no activity). Otherwise the number
        # of columns would depend on activity density rather than duration,
        # and a thread that ran longer but with sparse bursts would look
        # narrower than a dense short one. The frontend renders empty bins
        # with a "no activity" label.
        start_bin = _bin_start_epoch(start_epoch, unit)
        end_bin = _bin_start_epoch(end_epoch, unit)
        bins_out = []
        for bin_epoch in range(start_bin, end_bin + unit, unit):
            agents = bin_buckets.get(bin_epoch)
            members_by_key = bin_members.get(bin_epoch, {})
            agents_out = []
            if agents:
                for k, b in agents.items():
                    entry = {"key": k, "total": b.total}
                    members = members_by_key.get(k)
                    if members:
                        entry["members"] = sorted(members)
                    agents_out.append(entry)
                agents_out.sort(key=lambda a: (-a["total"], a["key"]))
            bins_out.append({"t": _iso_from_epoch(bin_epoch), "agents": agents_out})

        total_tokens = sum(a["total"] for bo in bins_out for a in bo["agents"])
        # Cost is computed from the same Buckets as tokens (one tier, one
        # formula) — the sum over buckets equals the thread's full cost,
        # just as total_tokens equals the sum of b.total.
        total_cost = round(sum(b.cost for agents in bin_buckets.values() for b in agents.values()), 2)
        # Distinct workflow runs in the session: the agentId of a subagent
        # launched within a workflow carries workflowRunId (see
        # tokens.walk); a regular thread without a workflow gives an empty
        # set -> 0.
        workflow_count = len({r["workflowRunId"] for r in recs if r["workflowRunId"]})

        git_info = git_events.scan_session(session_files.get(sid, []))

        threads.append(
            {
                "session": sid,
                "project": proj,
                # a human-readable name will be filled in by the service (src/service) later
                "title": sid,
                "start": _iso_from_epoch(start_epoch),
                "end": _iso_from_epoch(end_epoch),
                "durationSec": end_epoch - start_epoch,
                "totalTokens": total_tokens,
                "totalCost": total_cost,
                "workflowCount": workflow_count,
                "bins": bins_out,
                "cwd": git_info["cwd"],
                "gitBranch": git_info["gitBranch"],
                "events": git_info["events"],
            }
        )

    return {"threads": threads, "agentLabels": agent_labels}


def main():
    ap = tokens.JsonAwareParser()
    ap.add_argument("--json", action="store_true", help="print a single JSON object instead of a table")
    ap.add_argument("--limit", type=int, default=20, help="how many of the most recent sessions to take")
    ap.add_argument("--unit", type=int, required=True, help="bin size in seconds")
    ap.add_argument("--project", help="substring of the project path")
    ap.add_argument("--session", help="exact session id (single-session page); overrides recency selection")
    ap.add_argument("--group-workflows", action="store_true", help="merge agents of one workflow run into a single segment")
    a = ap.parse_args()

    if a.limit < 0:
        ap.error("--limit cannot be negative")
    if a.unit <= 0:
        ap.error("--unit must be a positive number of seconds")

    result = build_timeline(
        tokens.ROOT, limit=a.limit, unit=a.unit, project=a.project, session=a.session, group_workflows=a.group_workflows
    )
    out = {"schemaVersion": SCHEMA_VERSION, "unit": a.unit, **result}

    if a.json:
        print(json.dumps(out, ensure_ascii=False))
        return

    print(f"{'session':<10} {'project':<40} {'tokens':>10} {'sec':>8}")
    print("-" * 70)
    for t in result["threads"]:
        print(f"{t['session'][:8]:<10} {t['project'][:40]:<40} {t['totalTokens']:>10} {t['durationSec']:>8.0f}")


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        if "--json" in sys.argv:
            print(json.dumps({"error": str(e)}, ensure_ascii=False))
            sys.exit(1)
        raise

File: bb-plugin-token-usage-header/src/core/threads-timeline.ts
// Zod-based parsing of `tools/threads_timeline.py --json` output into a
// typed result, plus pure UI-layout helpers for the threads feed. No I/O
// here — the caller (src/service) runs the process; this module only ever
// sees a string. Never throws: malformed input from an external process is
// an expected case, not a bug, so every failure comes back as a tagged
// result the caller must handle explicitly. Mirrors src/core/parse.ts's
// contract for tools/tokens.py, but a separate schema/version — a different
// script, a different JSON shape.
import { z } from "zod";
import { gitEventSchema } from "./git-events";

/**
 * Version of the threads_timeline.py --json report format understood by
 * this bundle. Must match SCHEMA_VERSION in tools/threads_timeline.py — the
 * counter script is read from disk on every call, while this number lives
 * in the built bundle and only gets updated on rebuild. A separate version
 * from EXPECTED_SCHEMA_VERSION in src/core/types.ts (tools/tokens.py) —
 * different scripts, different contracts, the versions don't have to match.
 *
 * 1 -> 2: added the top-level agentLabels field (human-readable agent names
 * by key — see RawThreadsTimelineSchema below).
 * 2 -> 3: a thread gained totalCost (usage cost in USD at tokens.py's rate)
 * and workflowCount (number of distinct workflow runs in the session).
 * 3 -> 4: a bin's workflow segment (key == "workflow:<run>", only under
 * group_workflows) gained members — the real agentIds merged into the segment.
 * 4 -> 5: a thread gained cwd/gitBranch (its working directory and git
 * branch, read from the transcript) and events (pr/push facts from the same
 * transcript — see git_events.scan_session; commit events are appended
 * later, client-independent of the script, by
 * src/service/threads-timeline-service.ts's enrichCommits).
 */
export const EXPECTED_THREADS_TIMELINE_SCHEMA_VERSION = 5;

const AgentBinSchema = z
  .object({
    /** "main" for the main agent, otherwise the subagent's agentId — as in tokens.py --by agent. */
    key: z.string(),
    total: z.number().finite(),
    /**
     * Real agentIds merged into this segment — present ONLY when `key`
     * itself is a group workflow key (`workflow:<run>`, see
     * tools/threads_timeline.py::_bin_key under --group-workflows); for a
     * regular agent, key is already its real id and the field is absent.
     * Exists so the session chart (thread-chart.tsx) can highlight/dim a
     * segment by the selected agent — without it, which real agent
     * contributed within a merged segment would be indistinguishable.
     */
    members: z.array(z.string()).optional(),
  })
  .strict();

const TimelineBinSchema = z
  .object({
    /** ISO 8601 UTC start of the bin; a multiple of the report's `unit` seconds. */
    t: z.string(),
    agents: z.array(AgentBinSchema),
  })
  .strict();

// Shape exactly as threads_timeline.py --json prints one thread — the
// script doesn't know about BB projects at all, so this schema (used to
// validate its raw stdout below) never mentions bbProjectId/bbProjectName/
// threadId. Those three are attached by parseThreadsTimeline itself, right
// after this schema validates, with null defaults — see ThreadEntry below.
const RawThreadEntrySchema = z
  .object({
    session: z.string(),
    /** Raw transcript directory slug from python (`~/.claude/projects/<slug>`) — not to be confused with bbProjectId/bbProjectName below. */
    project: z.string(),
    /** Currently always equal to `session` — the service substitutes the human-readable name. */
    title: z.string(),
    start: z.string(),
    end: z.string(),
    durationSec: z.number().finite(),
    totalTokens: z.number().finite(),
    /** Cost of the thread's entire usage in USD, at the same rate as tokens.py (Bucket.cost). */
    totalCost: z.number().finite(),
    /** How many distinct workflow runs took part in the session (0 — a regular thread with no workflow). */
    workflowCount: z.number().int().nonnegative(),
    bins: z.array(TimelineBinSchema),
    /** The session's working directory, from its own transcript records — null when none carried both cwd and gitBranch. Used by the service layer to resolve commit events via a live `git log`; the parsed report otherwise only reads it, never runs anything with it. */
    cwd: z.string().nullable(),
    /** The session's git branch, same source/nullability as cwd. */
    gitBranch: z.string().nullable(),
    /** pr/push facts mined from the transcript by tools/git_events.py — commit events are appended on top by the service layer, not present in the script's own output. */
    events: z.array(gitEventSchema),
  })
  .strict();

/**
 * agentId (as in bins[].agents[].key, "main" for the main agent) ->
 * human-readable label. Built by threads_timeline.py from the subagent's
 * meta (description, else agentType, else the agentId itself as a fallback)
 * — see tools/threads_timeline.py::_agent_label. Top-level, not per-thread:
 * the same agentId across different threads of the slice shares one label.
 * An empty object is a valid case (a slice with no threads at all).
 */
const AgentLabelsSchema = z.record(z.string(), z.string());

const RawThreadsTimelineSchema = z
  .object({
    schemaVersion: z.number(),
    /** Bin size in seconds, as passed to --unit. */
    unit: z.number().finite(),
    threads: z.array(RawThreadEntrySchema),
    agentLabels: AgentLabelsSchema,
  })
  .strict();

export type AgentBin = z.infer<typeof AgentBinSchema>;
export type TimelineBin = z.infer<typeof TimelineBinSchema>;

/**
 * A thread as the rest of the plugin sees it — threads_timeline.py's own
 * fields plus the BB project/thread match added on top. bbProjectId/
 * bbProjectName/threadId/bbThreadTitle are always present (never optional)
 * but frequently null: a session that isn't tied to any BB thread (older
 * than threads-timeline-service.ts's scan window, or from outside BB
 * entirely) is the catch-all "Threads" bucket on the project picker, not a
 * parse failure. parseThreadsTimeline fills these four with null right
 * after validating the raw script output; src/service/threads-timeline-service.ts
 * overwrites them with a real match when it finds one. bbThreadTitle is the
 * BB thread's human-readable title (distinct from the script's own `title`
 * field above, which is always just the session id) — the UI prefers it as
 * the card's display name, falling back to a short session id when null.
 */
export type ThreadEntry = z.infer<typeof RawThreadEntrySchema> & {
  bbProjectId: string | null;
  bbProjectName: string | null;
  threadId: string | null;
  bbThreadTitle: string | null;
  /**
   * True when the matched BB thread is live (not archived). False for a
   * session with no BB thread match (the "Threads" bucket) — liveness is a
   * property of a BB thread, and there's none to read. Drives the feed's
   * green thread title.
   */
  isAlive: boolean;
  /**
   * True when work is happening in the matched BB thread right now — its main
   * turn is active or any background work runs (see deriveThreadLiveness).
   * Always false for an unmatched session. Drives the feed's blinking dot.
   */
  isWorking: boolean;
};

export type ThreadsTimeline = Omit<z.infer<typeof RawThreadsTimelineSchema>, "threads"> & {
  threads: ThreadEntry[];
};

export type ThreadsTimelineParseFailureReason =
  | "invalid_json"
  | "invalid_shape"
  | "script_error"
  | "schema_version_mismatch";

export interface ThreadsTimelineParseSuccess {
  ok: true;
  data: ThreadsTimeline;
}

export interface ThreadsTimelineParseFailure {
  ok: false;
  reason: ThreadsTimelineParseFailureReason;
  message: string;
}

export type ThreadsTimelineParseResult = ThreadsTimelineParseSuccess | ThreadsTimelineParseFailure;

export interface ThreadsTimelineScriptError {
  error: string;
}

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function fail(reason: ThreadsTimelineParseFailureReason, message: string): ThreadsTimelineParseFailure {
  return { ok: false, reason, message };
}

/** True when a parsed JSON value looks like threads_timeline.py's `{"error": "..."}` output. */
export function isThreadsTimelineScriptError(json: unknown): json is ThreadsTimelineScriptError {
  return isRecord(json) && typeof json.error === "string" && !("threads" in json);
}

/**
 * Parses the stdout of `threads_timeline.py --json`. Returns a tagged
 * result — never throws.
 */
export function parseThreadsTimeline(raw: string): ThreadsTimelineParseResult {
  const trimmed = raw.trim();
  if (trimmed.length === 0) {
    return fail("invalid_json", "empty output");
  }

  let json: unknown;
  try {
    json = JSON.parse(trimmed);
  } catch (e) {
    return fail("invalid_json", `not valid JSON: ${e instanceof Error ? e.message : String(e)}`);
  }

  // threads_timeline.py's own top-level exception handler prints
  // {"error": "..."} and nothing else — distinguish that from a malformed
  // real report before anything else looks at the shape.
  if (isThreadsTimelineScriptError(json)) {
    return fail("script_error", json.error);
  }

  if (!isRecord(json)) {
    return fail("invalid_shape", "expected a JSON object at the top level");
  }

  // Checked before threads/unit and anything else: a version mismatch means
  // the built plugin and tools/threads_timeline.py on disk speak different
  // dialects of the format, and the failure must name that instead of the
  // first data field the parser happens to reach (see
  // memory/decisions/token-usage-json-schema-version.md for the same
  // approach on the neighboring contract).
  if (json.schemaVersion !== EXPECTED_THREADS_TIMELINE_SCHEMA_VERSION) {
    const got =
      json.schemaVersion === undefined
        ? "the schema version field is missing"
        : `got version ${JSON.stringify(json.schemaVersion)}`;
    const remedy =
      typeof json.schemaVersion === "number" && json.schemaVersion > EXPECTED_THREADS_TIMELINE_SCHEMA_VERSION
        ? "Rebuild the plugin."
        : "Rebuilding the plugin won't help: update the plugin installation or check which tree tools/threads_timeline.py is being read from.";
    return fail(
      "schema_version_mismatch",
      `Plugin was built against a different version of the tools/threads_timeline.py counter: expected schema version ${EXPECTED_THREADS_TIMELINE_SCHEMA_VERSION}, ${got}. ${remedy}`,
    );
  }

  const result = RawThreadsTimelineSchema.safeParse(json);
  if (!result.success) {
    const message = result.error.issues.map((issue) => `${issue.path.join(".") || "<root>"}: ${issue.message}`).join("; ");
    return fail("invalid_shape", message);
  }

  // The raw schema validated the script's own fields; bbProjectId/
  // bbProjectName/threadId/bbThreadTitle aren't part of that contract (see
  // RawThreadEntrySchema's doc comment) — start every thread unmatched, for
  // the service layer to fill in when it enriches against bb.sdk.
  const data: ThreadsTimeline = {
    ...result.data,
    threads: result.data.threads.map((thread) => ({
      ...thread,
      bbProjectId: null,
      bbProjectName: null,
      threadId: null,
      bbThreadTitle: null,
      isAlive: false,
      isWorking: false,
    })),
  };

  return { ok: true, data };
}

// --- UI layout helpers -------------------------------------------------

/**
 * Each thread's width fraction relative to the longest by durationSec: the
 * longest gets 1.0, the rest are proportional to their own duration. Result
 * indexes match `threads`' indexes.
 *
 * Empty array -> empty array. When every thread has zero (or negative —
 * shouldn't happen, but isn't treated as an error here) duration, the
 * maximum is zero and there's nothing to divide by — every thread then gets
 * 1.0, not NaN/Infinity: single-point bars are drawn at the same minimal
 * width instead of disappearing.
 */
export function widthFractions(threads: readonly ThreadEntry[]): number[] {
  if (threads.length === 0) return [];
  const maxDuration = Math.max(...threads.map((t) => t.durationSec), 0);
  if (maxDuration <= 0) return threads.map(() => 1);
  return threads.map((t) => Math.max(t.durationSec, 0) / maxDuration);
}

/** Total usage of one bin — the sum of `total` across all agents within it. */
export function binTotal(bin: TimelineBin): number {
  return bin.agents.reduce((sum, agent) => sum + agent.total, 0);
}

// --- Thread liveness ---------------------------------------------------

/**
 * The BB-thread facts the feed's liveness indicators derive from, reduced to
 * exactly what {@link deriveThreadLiveness} needs. The service (imperative
 * shell) reads these off `bb.sdk.threads.list`'s response and hands them here;
 * this module never imports the SDK, so the mapping stays testable in
 * isolation — see memory/decisions/thread-liveness-signals.md.
 */
export interface ThreadLivenessInput {
  /** Epoch ms the BB thread was archived at, or null while it's still live. */
  archivedAt: number | null;
  /**
   * Epoch ms of the thread's most recent activity — the last transcript
   * record (`end`). The feed is a transcript snapshot, not a live turn feed:
   * bb.sdk.threads.list's own `status`/`runtime.displayStatus` describe the
   * *environment* lifecycle (provisioning/starting/stopping), not "an agent
   * turn is running now" — a local thread works at status "idle". So recency
   * of this record is the honest "working now" signal. NaN when unparseable.
   */
  lastActivityMs: number;
  /** Epoch ms "now" at snapshot time — injected, never read from the clock in core. */
  nowMs: number;
  /** How fresh lastActivityMs must be (ms) to count the thread as working. */
  workingWindowMs: number;
  /** Count of background work items running now (agents + commands + workflows + plan mode + goals). */
  activeWorkCount: number;
}

export interface ThreadLiveness {
  isAlive: boolean;
  isWorking: boolean;
}

/**
 * Derives the two feed indicators from a BB thread's raw liveness facts.
 * Pure and total: same input, same flags, no I/O.
 *
 * - `isAlive` — the thread is not archived.
 * - `isWorking` — the thread is alive AND either its last activity is within
 *   `workingWindowMs` of now, or some background work is running. An archived
 *   thread is never shown working, even with a stale flag (see the decision
 *   doc). A NaN `lastActivityMs` fails the recency test, never throws.
 */
export function deriveThreadLiveness(input: ThreadLivenessInput): ThreadLiveness {
  const isAlive = input.archivedAt === null;
  // No lower bound on the gap: a lastActivityMs slightly in the future (clock
  // skew) is "just now", still recent. NaN makes the comparison false.
  const recentlyActive = input.nowMs - input.lastActivityMs <= input.workingWindowMs;
  const isWorking = isAlive && (recentlyActive || input.activeWorkCount > 0);
  return { isAlive, isWorking };
}
